src/order/category/Preorder.lean

```diff
@@ -0,0 +1,77 @@
+/-
+Copyright (c) 2020 Johan Commelin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johan Commelin
+-/
+
+import category_theory.concrete_category
+import algebra.punit_instances
+
+/-! # Category of preorders -/
+
+open category_theory
+
+/-- The category of preorders. -/
+def Preorder := bundled preorder
+
+/-- Bundled monotone (aka, increasing) function -/
+structure preorder_hom (α β : Type*) [preorder α] [preorder β] :=
+(to_fun   : α → β)
+(monotone : monotone to_fun)
+
+namespace preorder_hom
+variables {α β γ : Type*} [preorder α] [preorder β] [preorder γ]
+
+instance : has_coe_to_fun (preorder_hom α β) :=
+{ F := λ f, α → β,
+  coe := preorder_hom.to_fun }
+
+@[ext] lemma ext (f g : preorder_hom α β) (h : ∀ a, f a = g a) : f = g :=
+by { cases f, cases g, congr, funext, exact h _ }
+
+lemma coe_inj (f g : preorder_hom α β) (h : (f : α → β) = g) : f = g :=
+by { ext, rw h }
+
+/-- The identity function as bundled monotone function. -/
+@[simps]
+def id : preorder_hom α α :=
+⟨id, monotone_id⟩
+
+instance : inhabited (preorder_hom α α) := ⟨id⟩
+
+@[simp] lemma coe_id : (@id α _ : α → α) = id := rfl
+
+/-- The composition of two bundled monotone functions. -/
+def comp (g : preorder_hom β γ) (f : preorder_hom α β) : preorder_hom α γ :=
+⟨g ∘ f, g.monotone.comp f.monotone⟩
+
+@[simp] lemma coe_comp (g : preorder_hom β γ) (f : preorder_hom α β) :
+  (g.comp f : α → γ) = g ∘ f := rfl
+
+@[simp] lemma comp_apply (g : preorder_hom β γ) (f : preorder_hom α β) (a : α) :
+  g.comp f a = g (f a) := rfl
```

Comment:
```suggestion
```

Basically, any definition written as a direct application of the constructor of `preorder_hom` can be marked as `@[simps]` and all its projections will be given a `simp` lemma to perform a iota-reduction automatically.